source-academy/sicp
# Generate JSON files from XML source files

# Description
This PR parses the source XML files into JSON files that can be easily parsed by interative-sicp on cadet-frontend. 

Currently the XML files have to be processed to remove unnecessary tags, generate ids, generate links and more. Preprocessing the files into JSON will minimise the amount of processing needed to be done on the frontend.

Three types of JSON files are generated:
1. toc.json
This file contains the table of contents, generated to work with [blueprint.js trees](https://blueprintjs.com/docs/#core/components/tree) without any modification.
2. toc-navigation.json
This file provides a mapping of current-page -> {prev-page, next-page}. This serves as a lookup table to determine which page to render when prev or next buttons are clicked.
3. Content
Each file containing the contents of the textbook is named xxx.json, where xxx is the name of the section or subsection.
The general format of the files are as follows:
```
[{
  tag: "Name of Node"
  body: "Contents of Node"
  ... // other attributes
  child: [{...}] // child nodes if any
},
{
   ...
}
]
```
Each object represents a node. Adjacent objects in an array represent sibling nodes (should be rendered one after the other). Nested objects (see child attribute) are objects that are nested within the current node. 

In general, the node's data is stored in the object, with formatting and stying left to the frontend. Documentation for the structure of each type of object and how the files are parsed will be added in the coming days.

## Implementation
The implementation closely follows the implementation of the web version of SICP. The main differences is that instead of pushing data to the `writeTo` array, data is added to javascript objects.

## Breaking Changes
No breaking changes are expected, since the changes are mostly additive.

## How to test
```
yarn clean
yarn json // (or yarn all)
```
This should generate the json files in the `/json/ `folder. 
```
yarn prepare
yarn try
```
This should move the json files into /docs_out/json/ and start a local web server. The files can be seen by going to `localhost:8080/json/xxx.json`, where `xxx` is the name of the section or subsection e.g. `localhost:8080/json/1.1.1.json`.

I have also tested the GitHub Actions deploy workflow. The files are currently hosted on my own GitHub page, and can be viewed at [https://www.samuelfangjw.com/sicp/json/1.1.1.json](https://www.samuelfangjw.com/sicp/json/1.1.1.json).

## Follow up and future plans
I will continue to touch up and simplify as needed. Documentation will also be added to the sicp wiki page.
